modules/phase_field/src/materials/PFParamsPolyFreeEnergy.C

```diff
@@ -0,0 +1,105 @@
+  params.addRequiredParam<Real>("int_width", "The interfacial width of void surface in the lengthscale of the problem");
+  params.addParam<Real>("length_scale", 1.0e-9, "defines the base length scale of the problem in m");
+  params.addParam<Real>("time_scale", 1.0e-9, "defines the base time scale of the problem");
+  MooseEnum poly_order("4 6 8");
+  params.addRequiredParam<MooseEnum>("polynomial_order", poly_order, "Order of polynomial free energy");
+  params.addRequiredParam<Real>("D0", "Diffusivity prefactor for vacancies in m^2/s");
+  params.addRequiredParam<Real>("Em", "Migration energy in eV");
+  params.addRequiredParam<Real>("Ef", "Formation energy in eV");
+  params.addRequiredParam<Real>("surface_energy", "Surface energy in J/m2");
+  return params;
+}
+
+PFParamsPolyFreeEnergy::PFParamsPolyFreeEnergy(const std::string & name,
+                             InputParameters parameters) :
+    Material(name, parameters),
+    _c(coupledValue("c")),
+    _T(coupledValue("T")),
+    _M(declareProperty<Real>("M")),
+    _grad_M(declareProperty<RealGradient>("grad_M")),
+    _kappa(declareProperty<Real>("kappa")),
+    _c_eq(declareProperty<Real>("c_eq")),
+    _W(declareProperty<Real>("barr_height")),
+    _Qstar(declareProperty<Real>("Qstar")),
+    _int_width(getParam<Real>("int_width")),
+    _length_scale(getParam<Real>("length_scale")),
+    _time_scale(getParam<Real>("time_scale")),
+    _order(getParam<MooseEnum>("polynomial_order")),
+    _D0(getParam<Real>("D0")),
+    _Em(getParam<Real>("Em")),
+    _Ef(getParam<Real>("Ef")),
+    _surface_energy(getParam<Real>("surface_energy")),
+    _JtoeV(6.24150974e18), // joule to eV conversion
+    _kb(8.617343e-5) // Boltzmann constant in eV/K
+{
+}
+
+void
+PFParamsPolyFreeEnergy::computeQpProperties()
+{
+  // Convert mobility from m^2/s to the length and time scale
+  Real D0_c = _D0*_time_scale/(_length_scale*_length_scale);
+
+  Real kbT = _kb*_T[_qp];
+
+  //Compute equilibrium concentration and diffusivity
+  _c_eq[_qp] = std::exp(-_Ef/kbT);
+  Real D = D0_c*std::exp(-_Em/kbT);
+
+  //Compute free energy integral constant, such that int^1_0 f_loc = KN*sqrt(W)
+  Real KN = 0.0;
+
+  switch (_order)
+  {
+    case 0: //Fourth oder
+      KN = 2.0/3.0;
+      break;
+    case 1: //Sixth order
+      KN = 3.0/16.0*std::sqrt(3.0) + 9.0/64.0*std::sqrt(2.0)*(std::log(-std::sqrt(2.0) + std::sqrt(3.0)) + std::log(std::sqrt(2.0) + std::sqrt(3.0)));
+      break;
+    case 2: //Eigth order
+      KN = 0.835510425;
+      break;
+    default:
+      mooseError("Error in PFParamsPolyFreeEnergy: incorrect polynomial order");
+  }
+
+  // Convert surface energy from J/m2 to eV/length_scale
+  Real surf_energy = _surface_energy*_JtoeV*_length_scale*_length_scale;
+
+  // Set interfacial parameter and energy barrier
+  _kappa[_qp] = surf_energy*_int_width/KN;
+  _W[_qp] = surf_energy/( 2.0*_int_width*KN );
+
+  Real Co = 0.0;
+  Real a = _c_eq[_qp];
+
+  switch (_order)
```

Comment:
I liked keeping the calculation of KN and Co separate, to keep the code easier to understand. It makes it easy to differentiate the various options for KN and the various options for Co.